src/utils/work/fetchWork.js

```diff
@@ -1,59 +1,51 @@
 import { Environment, isBrowser } from 'utils/deploymentInfo';
 
-import { getBackendStatus } from 'redux/selectors';
-
 import cache from 'utils/cache';
-import generateETag from 'utils/work/generateETag';
-import { dispatchWorkRequest, seekFromS3 } from 'utils/work/seekWorkResponse';
-
-// Temporarily using gene expression without local cache
-const fetchGeneExpressionWorkWithoutLocalCache = async (
-  experimentId,
+import dispatchWorkRequest from 'utils/work/dispatchWorkRequest';
+import downloadFromS3 from 'utils/work/downloadFromS3';
+import waitForWorkRequest from './waitForWorkRequest';
+
+// retrieveData will try to get the data for the given experimentId and ETag from
+// the fastest source possible. It will try to get the data in order from:
+// 1. Browser cache
+// 2. S3 (cache) via signedURL
+// 3. Worker
+//   3.1. Via socket (small data)
+//   3.2. Via S3 (large data) via signedURL
+const retrieveData = async (experimentId,
+  ETag,
+  signedUrl,
+  request,
   timeout,
   body,
-  backendStatus,
-  environment,
-  broadcast,
-  extras,
-  onETagGenerated,
   dispatch,
-  getState,
-) => {
-  // If new genes are needed, construct payload, try S3 for results,
-  // and send out to worker if there's a miss.
-  const { pipeline: { startDate: qcPipelineStartDate } } = backendStatus;
-
-  const ETag = await generateETag(
-    experimentId,
-    body,
-    extras,
-    qcPipelineStartDate,
-    environment,
-    dispatch,
-    getState,
-  );
+  useCache) => {
```

Comment:
In which cases is `useCache` false? The runClustering-like cases?

Reply:
`const useCache = body.name !== 'GeneExpression';` it is replacing the duplicated code for fetching genes vs the rest

Reply:
ah ok